### PR TITLE
Add Secp256k1 support

### PR DESCRIPTION
**Pull request by gtsonevv**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/near/near-api-js/blob/master/CONTRIBUTING.md).
- [x] Commit messages follow the [conventional commits](https://www.conventionalcommits.org/) spec
- [x] **If this is a code change**: I have written unit tests.
- [x] **If this changes code in a published package**: I have run `pnpm changeset` to create a `changeset` JSON document appropriate for this change.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

Adds support for secp256k1 curve

## Test Plan

All tests pass

## Related issues/PRs

https://github.com/near/near-api-js/pull/985
https://github.com/near/near-api-js/issues/1039
https://github.com/near/near-api-js/issues/440

### CODE REVIEW

**Comment by gagdiez** [comment]
I gave it a quick look, do we have any test against the chain using `secp256k1`? I think the current tests only look for internal consistency (i.e. what we signed, can be checked) but not necessarily for external consistency (what we signed, the chain can understand)

Making a small transfer after creating the account should be sufficient to prove correctness

**Review by vikinatora — APPROVED** [review]

**Review by andy-haynes — CHANGES_REQUESTED** [review]
Nice work!

I do think we should reexamine the original approach to avoid introducing unnecessary breaking changes and complexity in serialization, but I'm keen to hear other thoughts. If this does end up being the most desirable approach then at a minimum we should avoid breaking backward compatibility with missing class fields.

File: packages/crypto/package.json
Comment (gtsonevv) [reply packages/crypto/package.json:24]: not really, will fix it

File: packages/biometric-ed25519/src/index.ts
Comment (gtsonevv) [reply packages/biometric-ed25519/src/index.ts:21]: not really, I will fix it

File: packages/crypto/src/public_key.ts
Comment (gtsonevv) [reply packages/crypto/src/public_key.ts:24]: it will be removed

File: packages/transactions/src/schema.ts
Comment (gtsonevv) [reply packages/transactions/src/schema.ts:96]: The reason for this is that in newer versions of `borsh` it cannot be done as in https://github.com/near/near-api-js/pull/985. That's why we decided to remove these fields. https://github.com/near/borsh-js/issues/85

File: packages/crypto/package.json
Comment (andy-haynes) [reply packages/crypto/package.json:24]: Thanks for the explanation!

Let's stick with `secp256k1` for now and handle that as a follow-up task. It will be much easier to deal with the issues you mention after we have working modules anyway.

File: packages/transactions/src/schema.ts
Comment (andy-haynes) [reply packages/transactions/src/schema.ts:96]: Makes sense, thanks again for clarifying! I should have known nobody would add to the schema unless it was the only option 😄

File: packages/crypto/src/public_key.ts
Comment (andy-haynes) [inline packages/crypto/src/public_key.ts:32]: Since only one of the `this.*Key` properties will be set, could we implement getters for `data` and `keyType` on the initialized key pair so that we can continue to expose them?  e.g.
```ts
get keyPair() {
  return this.ed25519Key || this.secp256k1Key;
}

get keyType(): KeyType {
  return this.keyPair.keyType;
}

get data(): Uint8Array {
  return this.keyPair.data;
}

toString(): string {
  const encodedKey = baseEncode(this.data);
  return `${key_type_to_str(this.keyType)}:${encodedKey}`;
}
```

Curious to hear what you think. The borsh-TS mismatch makes it awkward to use these classes now that they're not tightly coupled to ed25519 😅

File: packages/crypto/src/public_key.ts
Comment (andy-haynes) [inline packages/crypto/src/public_key.ts:77]: Pretty sure this can be reduced to
```ts
if (decodedPublicKey.length !== keySize) { ... }
```

If `parts.length === 1`, then `keyType` was already determined from the length of the value. If an invalid length value (i.e. anything other than 64 or 32) were passed, `keyType` would be assumed to be ed25519 and fail because the length was not 32. Did I get that right?

File: packages/crypto/src/public_key.ts
Comment (andy-haynes) [inline packages/crypto/src/public_key.ts:102]: Getters for `data` and `keyType` would be useful here, too.

File: packages/transactions/src/signature.ts
Comment (andy-haynes) [inline packages/transactions/src/signature.ts:9]: The same getter pattern mentioned above would prevent breaking changes here, too.

**Review by andy-haynes — CHANGES_REQUESTED** [review]
Thank you for the clarifications on the implementation. Just a couple more things otherwise I think we're good to go 🚀

**Comment by andy-haynes** [comment]
@race-of-sloths 🦥

**Comment by race-of-sloths** [comment]
@gtsonevv Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
Are you going to win race-of-sloths this month? If so, you should speed up! Run, Sloth, run!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/gtsonevv?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/gtsonevv?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/gtsonevv)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/gtsonevv)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

**Comment by andy-haynes** [comment]
@race-of-sloths score 5

**Comment by race-of-sloths** [comment]
### 🌟 Score recorded!

@andy-haynes, thank you for scoring this pull request in the Race of Sloths!

File: packages/crypto/src/public_key.ts
Comment (gtsonevv) [reply packages/crypto/src/public_key.ts:32]: I agree, it will be much better with getters.

File: packages/crypto/src/public_key.ts
Comment (gtsonevv) [reply packages/crypto/src/public_key.ts:77]: Yep, you are right, I will fix it.

File: packages/crypto/test/key_pair.test.js
Comment (andy-haynes) [inline packages/crypto/test/key_pair.test.js:88]: These can be removed now that they're in the `describe` block above

**Review by andy-haynes — APPROVED** [review]
Awesome work! Just noticed two duplicated tests otherwise everything's good to go

**Comment by andy-haynes** [comment]
@gtsonevv would you mind rebasing this please? It should just be your work + release ahead of you in history.

**Comment by gtsonevv** [comment]
@andy-haynes rebase onto master?

**Review by andy-haynes — APPROVED** [review]

**Comment by race-of-sloths** [comment]
### ✅ PR is finalized! 
Your contribution is much appreciated with a final score of 5!
You have received 50 Sloth points for this contribution

**Pull request merged**
